src/units/abilities.cpp

```diff
@@ -476,6 +476,54 @@ bool unit::has_ability_type(const std::string& ability) const
 	return !abilities_.child_range(ability).empty();
 }
 
+//this string abilitie is used for purpose of add a second halo if condition is matched.
+//It is put here header declaration commodity.
+static void create_vector(std::vector<std::string>& image_list, const config& cfg, const std::string& tag_name)
+{
+	auto ret = std::find(image_list.begin(), image_list.end(), cfg[tag_name].str());
+	if(ret == image_list.end()){
+		image_list.push_back(cfg[tag_name].str());
+	}
+}
+
+const std::vector<std::string> unit::halo_or_icon_abilities(const std::string& tag_name, const map_location& loc) const
+{
+	std::vector<std::string> image_list;
+	for (const config::any_child sp : abilities_.all_children_range()){
+		bool is_active = ability_active(sp.key, sp.cfg, loc);
+		if(is_active && ability_affects_self(sp.key, sp.cfg, loc) && !(sp.cfg)[tag_name + "_image"].str().empty()){
+			create_vector(image_list, sp.cfg, tag_name + "_image");
+		}
+		if(is_active && !(sp.cfg)[tag_name + "_image_self"].str().empty()){
+			create_vector(image_list, sp.cfg, tag_name + "_image_self");
+		}
+	}
+
+	const unit_map& units = get_unit_map();
+
+	const auto adjacent = get_adjacent_tiles(loc);
+	for(unsigned i = 0; i < adjacent.size(); ++i) {
+		const unit_map::const_iterator it = units.find(adjacent[i]);
+		if (it == units.end() || it->incapacitated())
+			continue;
+		if ( &*it == this )
+			continue;
+		for(const config::any_child j : it->abilities_.all_children_range()) {
+			if(affects_side(j.cfg, side(), it->side())
+				&& it->ability_active(j.key, j.cfg, adjacent[i])
+				&& ability_affects_adjacent(j.key, j.cfg, i, loc, *it)
+				&& !(j.cfg)[tag_name + "_image"].str().empty())
+			{
+				create_vector(image_list, j.cfg, tag_name + "_image");
+			}
+		}
+	}
+	if(!image_list.empty()){
```

Comment:
Sorting is not needed if its size is less than 2, not if it's empty.

Reply:
fixed in https://github.com/wesnoth/wesnoth/pull/7856/commits/8b4b81f5eba56d4a9737b72578c74d065ba44324